docs/content/sips/008-using-oci-registries.md

```diff
@@ -0,0 +1,298 @@
+pushing it to the desired OCI registry:
+
+```bash
+$ spin bindle pull --bindle-server <server> --bindle-username <username> --bindle-password <password> [bindle-name]
+# then
+$ spin oci push
+```
+
+### Implementation
+
+A Spin application is made up of some metadata and component information
+together with the Wasm modules and static assets that made up those components.
+So conceptually, a Spin application is not *a single artifact*, but rather
+multiple distinct objects.
+
+This SIP proposes that, when distributed using an OCI registry, a Spin
+application would become a new OCI *artifact* with multiple layers
+(making the distinction clear, as *images* and *artifacts* are separate
+entities in OCI). Specifically, the media type used in this implementation
+for a Spin application is `application/vnd.fermyon.spin.application.v1+config`,
+and each Wasm module and static asset from the Spin components becomes an individual
+*layer* in the resulting registry entity. Because each file and Wasm module becomes
+a separate *layer*, we can efficiently de-duplicate and distribute applications.
+
+The remaining question is representing the Spin application definition — Spin
+introduced the internal representation of a *locked application* - an
+intermediate representation of a Spin application that *can* have a way to
+content-address Wasm modules and static assets. The implementation for this SIP
+uses the Spin locked application as the OCI configuration object.
+
+Below is an example of a Spin application with one component and one static
+asset, its OCI manifest, locked application, and resulting local directory
+structure when pulling the application locally.
+
+Consider the following `spin.toml` file for the application:
+
+```toml
+spin_version = "1"
+authors = ["Radu Matei <radu.matei@fermyon.com>"]
+description = ""
+name = "github-stars-webhook"
+trigger = { type = "http", base = "/" }
+version = "0.1.0"
+
+[[component]]
+id = "github-star-webhook"
+source = "target/spin-http-js.wasm"
+files = ["my-file.json"]
+allowed_http_hosts = ["https://hooks.slack.com"]
+[component.trigger]
+route = "/..."
+```
+
+Distributing it to the GitHub Container Registry:
+
+```bash
+$ spin oci push ghcr.io/radu-matei/spin-example:v1
+INFO spin_publish::oci::client: Pushed "https://ghcr.io/v2/radu-matei/spin-example/manifests/sha256:8f86a27fbc457416701c4d18680083f598076d0a52dca2a5936e92754a845ed1"
+
+$ spin oci pull ghcr.io/radu-matei/spin-example:v1
+INFO spin_publish::oci::client: Pulled ghcr.io/radu-matei/spin-example:v1@sha256:8f86a27fbc457416701c4d18680083f598076d0a52dca2a5936e92754a845ed1
+```
+
+This operation created the following local cache directory structure:
+
+```bash
+$ tree /Users/radu/Library/Application\ Support/fermyon/registry
+└── oci
+    ├── data
+    │   └── sha256:a4699e4f9ef3f4922f38f0d017aa26438908f38caf020a739e0ee27fe796eb02
+    ├── manifests
+    │   └── ghcr.io
```

Comment:
why have all of these subdirectories rather than just one subfolder to the manifest and config?

Reply:
One subfolder per what?
Per registry? Repository? Tag?

The directory structure makes it simple to directly access the desired file (manifest or config) in O(1).

If there are more manifests from multiple registries, multiple accounts under the same registry, or multiple repositories of the same account, they would share subdirectories.

The Wasm modules and data files on the other hand are all in flat directories, as they are always addressed by digest.

Which subdirectories would you suggest flattening exactly?

Reply:
I was wondering if it could be fully flattened as so:
```
ghcr.io-radu-matei-spin-example-v1
    │                   ├── config.json
    │                   └── manifest.json
```
But makes sense to navigate directories rather than try to predictably create a single directory path. I am converted to your way :)